src/clusterfuzz/_internal/bot/tasks/utasks/analyze_task.py

```diff
@@ -315,30 +322,79 @@ def execute_task(testcase_id, job_type):
+        error=ErrorType.NO_CRASH,
+        test_timeout=test_timeout,
+        job_type=job_type)
   # Update testcase crash parameters.
   update_testcase_after_crash(testcase, state, job_type, http_flag)
 
-  log_message = (
-      f'Testcase crashed in {crash_time} seconds (r{testcase.crash_revision})')
-  data_handler.update_testcase_comment(testcase, data_types.TaskState.FINISHED,
-                                       log_message)
-
   # See if we have to ignore this crash.
   if crash_analyzer.ignore_stacktrace(state.crash_stacktrace):
-    data_handler.close_invalid_uploaded_testcase(testcase, metadata,
-                                                 'Irrelavant')
-    return None
+    return uworker_io.UworkerOutput(
+        testcase=testcase, metadata=metadata, error=ErrorType.IGNORE_STACK)
+
+  test_for_reproducibility(testcase, testcase_file_path, state, test_timeout)
+  return uworker_io.UworkerOutput(
+      testcase=testcase,
+      metadata=metadata,
+      test_timeout=test_timeout,
+      crash_time=crash_time)
 
-  # Test for reproducibility.
+
+def test_for_reproducibility(testcase, testcase_file_path, state, test_timeout):
   one_time_crasher_flag = not testcase_manager.test_for_reproducibility(
       testcase.fuzzer_name, testcase.actual_fuzzer_name(), testcase_file_path,
       state.crash_type, state.crash_state, testcase.security_flag, test_timeout,
-      http_flag, gestures)
+      testcase.http_flag, testcase.gestures)
   testcase.one_time_crasher_flag = one_time_crasher_flag
 
+
+def utask_handle_errors(output):
+  """Handles errors that happened in utask_main."""
+  if output.error == ErrorType.BUILD_SETUP:
+    data_handler.update_testcase_comment(
+        output.testcase, data_types.TaskState.ERROR, 'Build setup failed')
+
+    if data_handler.is_first_retry_for_task(output.testcase):
+      build_fail_wait = output.uworker_env.get('FAIL_WAIT')
+      tasks.add_task(
+          'analyze',
+          output.uworker_input['testcase_id'],
+          output.job_type,
+          wait_time=build_fail_wait)
+    else:
+      data_handler.close_invalid_uploaded_testcase(
+          output.testcase, output.metadata, 'Build setup failed')
+  elif output.error == ErrorType.NO_CRASH:
+    handle_noncrash(output.testcase, output.metadata, output.testcase_id,
+                    output.job_type, output.test_timeout)
+  elif output.error == ErrorType.TESTCASE_SETUP:
+    # Unclear if this state is ever actually reached.
+    data_handler.update_testcase_comment(
+        output.testcase, data_types.TaskState.ERROR, 'Testcase setup failed')
+  elif output.error == ErrorType.IGNORE_STACK:
+    data_handler.close_invalid_uploaded_testcase(output.testcase,
+                                                 output.metadata, 'Irrelavant')
+
+
+def utask_postprocess(output):
+  """Trusted: Cleans up after a uworker execute_task, writing anything needed to
+  the db."""
+  # TODO(metzman): Get rid of this after main migration PR.
```

Comment:
Maybe be *slightly* more descriptive here?  ("Get rid of this after we have enabled untrusted variants for all Clusterfuzz tasks?")  (Just so that, if someone encounters this TODO years from now, they'll have a good idea of whether it can be deleted with confidence or not)

Reply:
It's just a note to myself to get rid of the testcase = output.testcase junk I did to avoid making this diff larger. I tried to make this diff small by leaving out any nonfunctional changes if possible.